### ISSUE TITLE
Session hung if server reachability changes from IPv6 to IPv4 on transport change

### ISSUE THREAD

**Issue by dave4445** [schwabe/ics-openvpn#761]
To make issues more manageable, I would appreciate if you fill out the details if applicable

# General information
1. Android Version
7.1.1
2. Android Vendor/Custom ROM
Sony
3. Device
Sony Xperia X Compact
4. Version of the app (version number/play store version/self built)
0.6.73 

# Description of the issue

If the peer vpn server has both A and AAAA dns entries and client can connect via either IPv4 or IPV6, client doesn't change between v6 and v4 on a transport change.

Specifically, if I connect on wifi when the wifi has both v4 and v6 internet access, openvpn app will connect via IPv6, however when wifi goes out of range and switch to LTE, if LTE only has v4 access, the VPN hangs and does not reconnect as it is still trying to reach the server via IPv6.

The client should change between v4 and v6 dynamically on access/reachability change.

once changed from wifi with v4/v6 to LTE with v4 only logs fill with:
 
2017-09-11 19:25:23 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:23 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:23 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:23 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:24 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:24 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:24 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:24 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:24 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:25 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:25 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:26 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:26 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:26 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:27 write UDP []: Network is unreachable (code=101)
2017-09-11 19:25:27 write UDP []: Network is unreachable (code=101)

**Comment by dave4445** [comment]
Note this can be worked around for now by adding "proto udp4" as a custom option which will force a IPv4 connection ignoring the AAAA dns entry of dual-stack vpn servers.

**Comment by zoifgu** [comment]
Forcing v4 seems simpler.  And simplifies adblocking.

**Comment by schwabe** [comment]
The problem is that OpenVPN tries to keep the connection established and just "floats" between different network. I have to check if there is an Android API that tells me about v4/v6 capability of networks

**Comment by dave4445** [comment]
Even without that, a repetitive unreachable error could trigger an automatic reconnect?

**Comment by schwabe** [comment]
Maybe. But OpenVPN does not do that. And adding that might be a problem too as people could just send faked ICMP unreachable messages and kill a VPN connection very easily by that.

**Comment by dimitris-personal** [comment]
Resurrecting this as I've been bit both by the same issue (mobile on IPv6 onto IPv4-only WLAN) then followed by its mirror image:  manual or long-timeout reconnect on that WLAN, then walk out of WLAN range and back to the mobile operator's effectively IPv6-only network (and back to network unreachable and need to manual reconnect or long timeout).

`proto udp4` would break connectivity (on float attempt) on NAT64/DNS64 mobile networks, until they also support 464XLAT (is it even a safe assumption that they all would?) __and__ the associated [android bug](https://issuetracker.google.com/issues/255040839) is fixed (hence "IPv6 only" above).  `proto udp6` would break things on IPv4 only WLANs.

would an option to take account of network unreachable errors for a reconnect __only for some period of time after a float has started__ be a reasonable compromise here?

A bit more stretch, if we can "remember" all the address records we received from DNS when the connection was started, would it be possible to then attempt a float first, instead of a reconnect, in that condition?  Basically extend the float logic to say "try with other addresses for same peer first"?

**Comment by dave4445** [comment]
This does really seem like a subset of a more generic situation.  A server you are connecting to may have multiple A or multple AAAA records and the client should be trying each one in turn (regardless of if v4 or v6), until a succesful connection is make.  If the connection drops/fails, the client should try a different IP from that DNS lookup.

If the client remembers which ones it was just connected to isn't probably important, trying amongst the DNS results (no matter how many of them it gets) is the important part.  This really isn't v4/v6 specific.  A client connecting to a server (or pool of servers) with multiple IPs behind the same DNS name should failover/hunt amongst them not just on startup but if one is unreachable or lost after connection.

**Comment by dimitris-personal** [comment]
@dave4445 you're right, going round-robin (or just randomly choosing)  seems far more robust than mixing in "special logic on unreachable error but only some of the time".

I also think that's still not going to solve 100% of the problem, specifically when initially connected over an IPv6-only NAT64/DNS64 network (likely a mobile operator), then moving to an IPv4 only WLAN.

If I'm not mistaken, when the original DNS request was made, the DNS64 component would have mapped any A records into IPv4-embedded-in-IPv6 AAAA records.  So when moving to the IPv4 WLAN we still don't have a reachable address in the mix.  I'd be happy if my assumption about DNS64 is wrong here.

Either way though round-robin-on-float would be a good option to have and a better place to be than where we are now.

**Comment by dave4445** [comment]
Yes, when switching DNS servers (due to technology/iface change) old DNS cache should be purged.  As you say, with the advent of DNS64, any DNS results from wifi should not be mixed with wan, etc...  This really isn't specific to openvpn, but more of an Android concern.  But as long as fresh DNS queries are made it shouldn't really have a practical issue.  Even if DNS64 results were used once for connectivity out the wrong interface the generated v6 connection would fail and app should fallback to the v4 result or vice-versa.

**Comment by dave4445** [comment]
I think the base issue here as @schwabe mentions is _when_ should openvpn give up and start over.  A timeout?  ENETUNREACH and friends? something else?  There's nothing wrong with sticking with a working connection, but when the server stops responding something should be done to try to reconnect.  At that decision point a peer IP change, peer family change, DNS lookup, etc.. can occur as normal.  From what I've seen openvpn does the right thing at connect time, it's just the dead peer issue we're dealing with.

**Comment by dimitris-personal** [comment]
There's obviously a signal, on network change, that makes the client initiate a float.  Here's what I'm thinking:

- When initiating a float, keep and re-use the set of IP addresses from the original resolution in any manner, round-robin or random, doesn't matter much.

meanwhile, and I realize this is potentially much more of a stretch goal, at the same time when we initiate the float:

- In parallel with the ongoing float effort, stop any already-pending DNS request and start a new one to get a fresh, new-network-appropriate, set of address records.  If/when this is resolved replace the address bag with this one.  This might need to be on some exponential backoff type schedule while the float is still in progress.
- If the float has already succeeded do nothing else, but if not, start using the new address bag for the float.

This should give us close to best of all worlds:  Reuse connection where appropriate, and transparently reconnect when float isn't possible.

**Comment by dimitris-personal** [comment]
There's also [this](https://github.com/schwabe/ics-openvpn/issues/1299#issuecomment-802780430) which suggests that it's possible to disable float and force reconnects on all network changes?  I'm not sure if I'm reading that right though and if so how to enable that behavior.

> the client does not need to reconnect if the server uses peer-id

How do I disable peer-id?

**Comment by schwabe** [comment]
peer-id cannot be really disabled. All OpenVPN servers that are at least 2.4 or newer automatically use peer-id. I need to look into this situation more if the Android API can help me with this situation or not or if need to allow reconnecting on every network change.

**Comment by dimitris-personal** [comment]
Just a quick note that the [Android CLAT issue](https://issuetracker.google.com/issues/255040839) appears to be fixed in the February update of A13 (Pixel 7).  Means I can use IPv4 address literals and "safely" move between Google Fi mobile data and WLANs, some of the latter only having IPv4 connectivity.

Still can't use server hostname(s) though, as the server is reachable over both IPv4 and IPv6, as discussed above.
